### PR TITLE
Addition of a connectivity check module

### PR DESCRIPTION
**Pull request by paribeiro10**
Addition of a connectivity check module meant to handle connectivity changes, plus the respective sample screen to showcase the feature. This is usually a common class replicated in many projects, meaning it qualifies for ADAL's acceleration purpose.

It makes use of a Broadcast Receiver for the `android.net.conn.CONNECTIVITY_CHANGE` filter. Furthermore, it has a utility class to query the connectivity type (WIFI, MOBILE) and if there is a connection available (true or false). All actions are logged to log cat when debugging as well.

### CODE REVIEW

**Comment by extmkv** [comment]
Maybe we can merge this module with the network module to avoid one more module.

What do you think about that? @jzeferino @paribeiro10 @NunoSSilva @PedroOkawa @Andrer757

**Comment by jzeferino** [comment]
@extmkv It make sense to be a separated module since we can use it witout retrofit/network module.Yes it will be another module, but IMO its not a biggie. Regarding the code, will review it tomorow. @paribeiro10. Another thing @paribeiro10 try to avoid the separated commits regarding checkstyle/pmd, etc. thanks for the contrib.

**Comment by paribeiro10** [comment]
@jzeferino I will be more careful next time, thanks.
@extmkv I agree with @jzeferino on the separate module option. My understanding was that the adal-network handled API calls and this just "observes" device connectivity.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/ConnectionChangeReceiver.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.IntentFilter;
+import android.util.Log;
+
+/**
+ * Abstract class meant to offer connectivity change detection capabilities,
+ * using the Android system's connection change system filter.
+ */
+@SuppressWarnings("unused") public abstract class ConnectionChangeReceiver extends BroadcastReceiver {
```
Comment (jzeferino) [inline adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/ConnectionChangeReceiver.java]: @paribeiro10 the `supress` is really needed here? There is another way to solve this other than supress?

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package com.massivedisaster.adal.connectivity;
+
+/** Utility class meant to hold all constants for the Network sub module. */
+@SuppressWarnings({"unused", "WeakerAccess"}) public final class NetworkConstants {
+
+    /** ########## Misc. ########## */
```
Comment (jzeferino) [inline adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: @paribeiro10 We don't use this type of comment in ADAL instance members. Use only the message. Example here: https://github.com/massivedisaster/ADAL/blob/master/adal-application-state/src/main/java/com/massivedisaster/adal/applicationstate/ApplicationStateManager.java#L38

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    public static final String LOG_TAG = "CADL-Network";
+
+    /** ########## Exception Messages ########## */
```
Comment (jzeferino) [inline adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: Same as https://github.com/massivedisaster/ADAL/pull/125/files#r137398680

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    public static final String ASSERTION_ERROR = "Instantiating utility class.";
+
+    /** ########## Error Messages ########## */
```
Comment (jzeferino) [inline adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: Same as https://github.com/massivedisaster/ADAL/pull/125/files#r137398680

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    public static final String INVALID_CONTEXT_INSTANCE = "::Invalid Context instance: ";
+
+    /** ########## Information Messages ########## */
```
Comment (jzeferino) [inline adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: Same as https://github.com/massivedisaster/ADAL/pull/125/files#r137398680

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    public static final String CONNECTION_CHANGE_UNREGISTER_NOK = "::Unable to unregister receiver.";
+    public static final String CONNECTION_CHANGE_UNREGISTER_OK = "::Unregistered receiver.";
+    public static final String CONNECTION_CHANGE_REGISTER_OK = "::Registered receiver.";
+
+    /** Instantiates a new NetworkConstants. Private to prevent instantiation. */
+    private NetworkConstants() {
+        throw new AssertionError(ASSERTION_ERROR); // Throw an exception if this *is* ever called
```
Comment (jzeferino) [inline adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: @paribeiro10 This comment could be removed, otherwise, place it above the throw.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkUtils.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+package com.massivedisaster.adal.connectivity;
+
+import android.content.Context;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+
+/** Utility class for the Network sub module. */
+@SuppressWarnings({"unused", "WeakerAccess"}) public final class NetworkUtils {
```
Comment (jzeferino) [inline adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkUtils.java]: Same as https://github.com/massivedisaster/ADAL/pull/125/files#r137397807

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkUtils.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+    public static final int TYPE_NOT_CONNECTED = 0;
+    public static final int TYPE_WIFI = 1;
+    public static final int TYPE_MOBILE = 2;
+
+    /** Instantiates a new NetworkUtils. Private to prevent instantiation. */
+    private NetworkUtils() {
+        // Throw an exception if this *is* ever called
```
Comment (jzeferino) [inline adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkUtils.java]: @paribeiro10 as opposed to https://github.com/massivedisaster/ADAL/pull/125/files#r137399198 the comment is present and above the throw.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/ConnectionChangeReceiver.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+    public static final String CONNECTIVITY_CHANGE_FILTER = "android.net.conn.CONNECTIVITY_CHANGE";
+
+    private Context mContext;
+
+    /**
+     * Procedure meant to handle this {@link BroadcastReceiver}'s registration to any {@link Activity} instance.
```
Comment (jzeferino) [inline adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/ConnectionChangeReceiver.java]: @paribeiro10 The way the method documentation start ins't the way ADAL has other methods comments.
Here is an example: https://github.com/massivedisaster/ADAL/blob/master/adal-alarm/src/main/java/com/massivedisaster/adal/alarm/AlarmManager.java#L81

Every method should have a short description, describing waht it does, like: "Cancel alarm added before".

Please take a look at all method documentation started by "Procedure".

**Comment by jzeferino** [comment]
@paribeiro10 It could be done in this PR or in a new one, but please don't forget to add the docs about the sample.

Take a look at the [docs](https://github.com/massivedisaster/ADAL/tree/master/docs to create) for existing modules. 

Don't forget to later add the doc in the readme [Usage](https://github.com/massivedisaster/ADAL#usage) section

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/ConnectionChangeReceiver.java (hunk 1 above)
Comment (paribeiro10) [reply adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/ConnectionChangeReceiver.java]: In this case, the `suppress` is unneeded.
I have the habit of adding these 'suppress' when developing to remove the lint warnings from Android Studio (when you create the class it is not used as of yet).
I will remove it.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java (hunk 2 above)
Comment (paribeiro10) [reply adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: Duly noted. I will also change this.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java (hunk 3 above)
Comment (paribeiro10) [reply adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: Duly noted. I will also change this.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java (hunk 4 above)
Comment (paribeiro10) [reply adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: Duly noted. I will also change this.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java (hunk 5 above)
Comment (paribeiro10) [reply adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: Duly noted. I will also change this.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java (hunk 6 above)
Comment (paribeiro10) [reply adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: Duly noted. I will also change this.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkUtils.java (hunk 7 above)
Comment (paribeiro10) [reply adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkUtils.java]: Same as https://github.com/massivedisaster/ADAL/pull/125#discussion_r137652941

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkUtils.java (hunk 8 above)
Comment (paribeiro10) [reply adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkUtils.java]: I removed it as it is not needed. Well spotted on that one.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/ConnectionChangeReceiver.java (hunk 9 above)
Comment (paribeiro10) [reply adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/ConnectionChangeReceiver.java]: Duly noted. I will also change this.

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package com.massivedisaster.adal.connectivity;
+
+/**
+ * Utility class meant to hold all constants for the Network sub module.
+ */
+@SuppressWarnings({"unused", "WeakerAccess"}) public final class NetworkConstants {
```
Comment (jzeferino) [inline adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: @paribeiro10 forgot this one?

File: docs/adal-connectivity.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
+# adal-connectivity
+Android library to verify if the device has an Internet connection and what it's type is!
```
Comment (jzeferino) [inline docs/adal-connectivity.md]: @paribeiro10 replacing "and what it's type is!" with "and what type it is" seems more clear to read.

File: docs/adal-connectivity.md [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+### Download
+
+Gradle:
+
+```gradle
+dependencies {
+  compile 'com.massivedisaster.adal:adal-connectivity:0.1.15'
+}
+```
+### Usage
```
Comment (jzeferino) [inline docs/adal-connectivity.md:13]: @paribeiro10 the Usage should have activity/fragment/application/service code for the user to understand how to call this utility methods. I suggest a more simple usage such as calling the `checkConnectivity` and show its result in a `Toast`, for example.

File: sample/src/main/java/com/massivedisaster/adal/sample/feature/connectivity/FragmentConnectivityAware.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            handleConnectivityStatusChange(isOnline);
+        }
+    }
+
+    /**
+     * Log the device Internet connectivity status.
+     *
+     * @param isOnline boolean value indicating whether the device has a connection established or not.
+     */
+    private void handleConnectivityStatusChange(final boolean isOnline) {
```
Comment (jzeferino) [inline sample/src/main/java/com/massivedisaster/adal/sample/feature/connectivity/FragmentConnectivityAware.java:132]: @paribeiro10 This one its just a enhancement. You could remove the "duplicated" lines here, adding a ternary operator to get the string: 

```
String deviceConnectivity = isOnline ? getString(R.string.connectivity_device_online) : getString(R.string.connectivity_device_offline);

mTxtMessage.setText(Connectivity);
Log.d(getActivity().getClass().getName(), Connectivity);
```

Plus, the `Log` here seems irrelevant.

File: sample/src/main/res/layout/fragment_connectivity_aware.xml [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    android:orientation="vertical">
+
+    <TextView
+        android:id="@+id/txtMessage"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_centerInParent="true"
+        android:textColor="@color/colorPrimary"
+        android:textSize="22sp"
+        android:visibility="visible"/>
```
Comment (jzeferino) [inline sample/src/main/res/layout/fragment_connectivity_aware.xml]: @paribeiro10 `android:visibility="visible"` Irrelevant.

**Review by jzeferino — CHANGES_REQUESTED** [review]

File: adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java (hunk 10 above)
Comment (paribeiro10) [reply adal-connectivity/src/main/java/com/massivedisaster/adal/connectivity/NetworkConstants.java]: Duly noted. Removed.

File: docs/adal-connectivity.md (hunk 11 above)
Comment (paribeiro10) [reply docs/adal-connectivity.md]: Duly noted. Replaced.

File: docs/adal-connectivity.md (hunk 12 above)
Comment (paribeiro10) [reply docs/adal-connectivity.md:13]: Made changes to the class to make it more clear for this purpose.
I can remove the `TextView` and leave the Toast only. Please advise.

File: sample/src/main/java/com/massivedisaster/adal/sample/feature/connectivity/FragmentConnectivityAware.java (hunk 13 above)
Comment (paribeiro10) [reply sample/src/main/java/com/massivedisaster/adal/sample/feature/connectivity/FragmentConnectivityAware.java:132]: Good catch. Replaced.

File: sample/src/main/res/layout/fragment_connectivity_aware.xml (hunk 14 above)
Comment (paribeiro10) [reply sample/src/main/res/layout/fragment_connectivity_aware.xml]: Removed.

File: docs/adal-connectivity.md (hunk 12 above)
Comment (jzeferino) [reply docs/adal-connectivity.md:13]: Your call.

**Comment by jzeferino** [comment]
@paribeiro10 why did you changed the original travis configuration?

**Comment by paribeiro10** [comment]
@jzeferino I had a brain freeze last Tuesday. I removed the filler commits and corrected the issues.
The Travis configuration should be fine now. I changed it because I can't run the check plugin on my Windows machine... It got changed because of merge mix-ups.
